lang/funcs/collection.go

```diff
@@ -800,7 +800,13 @@ var MatchkeysFunc = function.New(&function.Spec{
 		},
 	},
 	Type: func(args []cty.Value) (cty.Type, error) {
-		if !args[1].Type().Equals(args[2].Type()) {
+		argTypes := make([]cty.Type, 2)
+		for i := 0; i < 2; i++ {
```

Comment:
Looking at the old logic here, I think the intent of only comparing the types of arguments 1 and 2 was to allow using matches of one type to produce another type.

For (contrived) example:

```hcl
matchkeys(
  [
    {foo = "bar"},
    {foo = "baz"},
    {foo = "beep"},
  ],
  ["a", "b", "c"],
  ["a", "c"],
)
```

If I understand the behavior of the function correctly, previously that would've produced the following result:

```hcl
[
    {foo = "bar"},
    {foo = "beep"},
]
```

So with that said, I think perhaps this should still only include arguments 1 and 2 in the unification, and leave argument 0 to just decide the result type. The error message before was implying that all three must be the same, but I think it's the error message that was wrong here... it should've been something like "key list and search list must have the same element type".

What do you think? :thinking:

Reply:
It is only including args 1 and 2 (`args[i+1]`); I will make that more clear:

https://github.com/hashicorp/terraform/blob/f2a14d7c183685c6ae71b70fc34285a9f8f0c47a/lang/funcs/collection.go#L804-L806

And I agree on the error message, I noticed it too then promptly forgot - thank you!

Reply:
Ahh, sorry... that `+ 1` was too subtle for my eye to catch it. :roll_eyes: 

Given this, maybe worth adding a comment above this loop explaining that it's collecting the types of arguments 1 and 2 for unification, though honestly I kinda feel inclined to just write these references out directly so the intent is more explicit:

```go
ty, _ := convert.UnifyUnsafe([]cty.Type{args[1].Type(), args[2].Type()})
```

I don't feel particularly strongly about any of this; just trying to put myself in the shoes of a hypothetical future reader of this.

Reply:
I like that a _lot_ better; updated.